src/Dhall/Core.hs

```diff
@@ -1396,6 +1429,15 @@ isNormalized e = case shift 0 "_" e of  -- `shift` is a hack to delete `Note`
         App NaturalIsZero (NaturalLit _) -> False
         App NaturalEven (NaturalLit _) -> False
         App NaturalOdd (NaturalLit _) -> False
+        App (App OptionalBuild t) k0 -> isNormalized t && isNormalized k0 && not (check0 k0)
+          where
+            check0 (Lam _ _ (Lam just _ (Lam nothing _ k))) = check1 just nothing k
+            check0 _ = False
+
+            check1 just nothing (App (Var (V just' n)) _) =
+                just == just' && n == (if just == nothing then 1 else 0)
```

Comment:
Related to your previous question: this is very similar to how the more efficient version of normalization would work.  It's just that I wrote `isNormalized` long after I wrote `normalize` so I wrote `isNormalized` with more careful attention to performance